### ISSUE TITLE
DSL

### ISSUE THREAD

**Issue by tessi** [bitcrowd/sshkit.ex#3]
The `dsl` branch has a rough outline of what the high-level API could look like.

It should allow you to specify

1. the hosts to run on
2. the context for the command/operation
  - working directory
  - user
  - group
  - umask
  - env
3. and then run this command on all hosts in the context
  - sequentially
  - in parallel
  - in groups of given size

A rough draft can be found in `sshkit.ex` @doc comments (on the `dsl` branch).

Open questions/things to do:

- How do we cache connections?
  - should be a KV-store erlang process
  - probably store connections by `{hostname, port, username}` as the key?
  - commands always go to this process to fetch a connection
    - if a connection with the same key doesn't exist yet, create, cache and return it
    - else: return the cached connection unless it timed-out
  - so it's more of a "connection provider" than just a "cache"
- How do we reuse the cached connections?
  - do we pass them in to the `SSHKit.run`  and `.upload` functions?
  - do these methods access the (global) cache process?

Thoughts:

- use `GenServer` for the connection provider
- could be set as part of the context, but have a default value so you do not need to explicitly pass it when just using the DSL

```elixir
context = SSHKit.context(hosts, provider: SSHKit.ConnectionProvider.new)

# in SSHKit

  def context(hosts, options \\ [provider: SSHit.ConnectionProvider.default])
    …
  end

  def run(context, command) do
    # fetch connections from the context's connection provider
    # execute
  end
```

**Comment by pmeinhardt** [comment]
I've split this up into smaller, more manageable tasks (see references listed above).
